### DISCUSSION TITLE
Clarification Required on the solve_linear Method in DAFoamSolver within DAFOAM Source Code

### DISCUSSION THREAD

**Discussion by lnhang0618** [mdolab/dafoam#492] [General]
Hello Professor He Ping,

I hope this message finds you well. I have been reviewing the DAFOAM source code and have come across a snippet in the **solve_linear** method within DAFoamSolver that has left me puzzled.

Here is the portion of the code I am referring to:

```
# right hand side array from d_outputs
dFdWArray = d_outputs["%s_states" % self.discipline]

```
I am struggling to comprehend the exact functioning of this part. Based on my understanding of the **apply_linear** definition, it seems that **d_output["aero_states"]** should be equivalent to **dRdw ^T * d_residua**l, implying that it represents **dw** in rev mode . This leads me to question why, in this scenario, **dFdW is** equated to **dw**. Is this a correct interpretation, or am I missing something in my understanding?

**Comment by lnhang0618** [comment]
But the function looks like a process in paper.
![1695710249184](https://github.com/mdolab/dafoam/assets/77599661/ccf96725-6dc0-4d48-826d-1060076751cf)

**Comment by friedenhe** [comment]
Your understanding of how the apply_linear function works is correct. This function computes the matrix-vector products for dRdInput^T*d_residual["states"].  However, for the solver_linear function, the `d_ouputs` dict saves the dFdW vector instead of the seed for matrix-vector products. The solve_linear solves the adjoint equation and output the adjoint vector psi, instead of matrix-vector products. So, apply_linear and solve_linear have very different input/output structures. solver_linear corresponds to the block 5  and apply_linear corresponds to the block 7 in the above diagram.
